mix.exs

```diff
@@ -43,7 +43,9 @@ defmodule GSearcher.MixProject do
       {:wallaby, "~> 0.26.2", [only: :test, runtime: false]},
       {:sobelow, "~> 0.10.4", [only: [:dev, :test], runtime: false]},
       {:mox, "~> 1.0.0", [only: :test]},
+      {:mimic, "~> 1.3", only: :test},
```

Comment:
Normally I prefer using Mox over Mimic because Mimic is really flexible, it allows us to mock everything, example I'm not sure about this case

```
use ExUnit.Case, async: true
use Mimic

test "invokes add once and mult twice" do
  Calculator
  |> stub(:add, fn x, y -> :stub end)
  |> expect(:add, fn x, y -> x + y end)
  |> expect(:mult, 2, fn x, y -> x * y end)

  assert Calculator.add(2, 3) == 5
  assert Calculator.mult(2, 3) == 6

  assert Calculator.add(2, 3) == :stub
end
```

we might return the method `add` in the Calculator but the test still passed because the Mimic stubted that method 💥 ,

Reply:
In Mimic, the `stub method` means it will always return the value no matter how many times you call it, the only exception is when you define an `expect`, it will return that once and will take priority over `stub`. If you call the second or any number of times after, it will return the `stub` value which they show in the example you gave 🙂 

On Mox, it seems more for certain use cases because it requires anything implementing a certain behavior, like client behaviors when calling an external API, GenServers, etc. 

This might be great for being explicit, but we can do the same thing in Mimic without having to define any callbacks. It actually implements a lot of the same features because it took inspiration from Mox and was made after Mox.

The problem here was that I needed Mimic to implement a test for `fail to save user` it will `redirect to the homepage with an error`. With Mox, I would need to abstract some behavior from either the `Accounts` or `Repo` which doesn't make much sense, but with Mimic, we can just `copy` the module we want to stub, and we can do the same expectation. There's also a `verify!()` method similar to Mox to ensure the function is called with the right arguments.

I'm not really sure why Mox is better over Mimic, from reading Mox's [blog example](https://dashbit.co/blog/mocks-and-explicit-contracts), they have something like 
```
[MyApp] -> [HTTP Client] -> [Twitter API]
```
and their solution is to create a `MyApp.Twitter Behavior` to have the flow be
```
[MyApp] -> [MyApp.Twitter (contract)]
[MyApp.Twitter.HTTP (contract impl)] -> [HTTPClient] -> [Twitter API]
```
and stub the `MyApp.Twitter.HTTP` to their own thing, like `MyApp.Twitter.InMemory` (The Mock)
We'd still be defining the expectation in the test. 🤔 

But with Mimic, we'd do something like
```
[MyApp] -> [MyApp.Twitter.HTTP] (Stub this one) -> [HTTPClient] -> [Twitter API]
```
It seems to me like they both solve the same issue but Mimic makes it easier, what do you think?

Reply:
I never use Mimic yet, but could you try to stub or expect a non-existing method with Mimic, does that raise any error?

like adding the

```
expect(Repo, :non_existing_method, fn _ -> {:error, %Ecto.Changeset{valid?: false}} end)
```

into https://github.com/bterone/g-searcher/blob/feature/oauth-signup/test/g_searcher_web/controllers/authentication_controller_test.exs#L49

When we run that test, does the test still passed or not?

I guess no because we had the `verify!` method there?

Reply:
I tested it, it seems to work as expected

![image](https://user-images.githubusercontent.com/11751745/99636794-81601f80-2a76-11eb-91d9-5ffebd47437a.png)